### DISCUSSION TITLE
Multiple Versions of Same Flow with different Parameters

### DISCUSSION THREAD

**Discussion by LINKDHS-ITS** [PrefectHQ/prefect#4543] [General]
I currently have a scheduling system that is organized by different "jobs". These jobs are essentially python classes that can be instantiated with varying parameters. For example, I have a job class that moves files from an SFTP to a local directory and I create multiple instances of it through the UI and specify the parameters through the UI as well without creating additional code.

Is it possible to do something similar with Prefect? Part of the advantage of the above system is that I can have non-technical people add jobs without adding any code or doing a release. 

Perhaps I'm not thinking about Prefect's tasks and flows correctly, but any hep or clarity would be appreciated.

**Comment by cicdw** [comment]
Hi @linkdkm - you can absolutely achieve this; Prefect has a concept of "Parameter" that allows you to specify pieces of configuration at run time (or even [schedule these values to change](https://docs.prefect.io/core/concepts/schedules.html#varying-parameter-values) if your work is on a schedule!) from the UI (or the API).  

However, Prefect does not currently let you alter the structure of your Flow from the UI so you'll need to account for that in your design -- most users use some combination of Parameters that affect what tasks run / how they run, along with Prefect's native [conditional logic](https://docs.prefect.io/core/idioms/conditional.html)

**Reply by LINKDHS-ITS** [reply]
Hi Chris, thanks for the reply. I actually asked the same question in the community slack and @kvnkho was kind enough to point me in a similar direction. 

Unfortunately, I don't think that the current functionality is quite what I'm looking for, though I do like that you can easily set up multiple schedules with various parameters. 

For my use cases, there would ideally be a drop down of registered flows, almost like templates, and then you could specify the unique parameters and schedules for that particular instance of that flow. That way you could manage and monitor it independently without having to write any code, it would just make an entry to the datastore. Alternatively, and likely easier to implement, would be a "Copy Flow" button, probably placed near the current Delete button. I spent some time experimenting and I think that it could potentially work. Below is some code that tests the concept, though I know if this were to ever be implemented, it would use the API directly. Essentially it just creates a new Flow with the same parameters and swaps out the "name". I wouldn't be surprised if this could lead to issues, but it worked for the few minutes that I tested it. 

Also, for your reference, Kevin mentioned he would note the question/request since it has come up before.

    `
    import prefect
    from prefect import task, Flow, Parameter
    
    @task
    def hello_task(param):
        logger = prefect.context.get("logger")
        logger.info(f"Hello world! - says {param}")
    
    with Flow("say_hello") as flow:
        name = Parameter("name")
        hello_task(name)
    
    class FlowCopy(Flow):
        def __init__(self, new_name=None, **kwargs):
            kwargs['name'] = new_name
    
            del kwargs['_cache']
            del kwargs['logger']
            del kwargs['_slug_counters']
            del kwargs['slugs']
            del kwargs['constants']
            del kwargs['_reference_tasks']
            del kwargs['_prefect_version']
            
            super().__init__(**kwargs)
    
    flow2 = FlowCopy(new_name='say_hello_copy', **vars(flow))
`

**Reply by cicdw** [reply]
Interesting -- any reason you can't register multiple Flows?  Your description makes sense, but ultimately you'll want to identify which template was used for each run, and that's all achievable through multiple Flows

**Reply by LINKDHS-ITS** [reply]
You're right that registering multiple Flows would be the most straightforward and wouldn't be difficult. However, we're a small non-tech company with 1.5 developers (I'm an Ops guy that picked up coding) and so most of our end users who would managing the flows would be non-developers. So the more that the non-devs can can do in the UI, the less the devs have to be responsible for. 

At the same time, I think my perspective is also influenced by our current business logic which could possibly be adapted to better fit Prefects paradigms.
